aeg/src/aeg.rs

```diff
@@ -59,19 +69,58 @@ impl From<&Program> for AbstractEventGraph {
 }
 
 impl AbstractEventGraph {
-    pub fn critical_cycles(&self) -> Vec<Vec<NodeIndex>> {
-        critical_cycles(&self.graph)
+    /// Find all neighbors of a node, taking into account transitive po edges
+    pub fn neighbors(&self, node: NodeIndex) -> Vec<NodeIndex> {
+        let mut po_neighbors = self.transitive_po_neighbors(node);
+        let mut close_non_po_neighbors: Vec<NodeIndex> = self
+            .graph
+            .edges(node)
+            .filter_map(|edge| match edge.weight() {
+                AegEdge::ProgramOrder => None,
+                AegEdge::Competing => Some(edge.target()),
+            })
+            .collect();
+
+        close_non_po_neighbors.append(&mut po_neighbors);
+        close_non_po_neighbors
+    }
+
+    fn transitive_po_neighbors(&self, node: NodeIndex) -> Vec<NodeIndex> {
+        // find all the po neighbors of this node, and all the po neighbors of them
+        let close_po_neighbors = self
+            .graph
+            .edges(node)
+            .filter_map(|edge| match edge.weight() {
+                AegEdge::ProgramOrder => Some(edge.target()),
+                AegEdge::Competing => None,
+            });
+
+        let mut neighbors: Vec<NodeIndex> = close_po_neighbors.clone().collect();
+
+        for n in close_po_neighbors {
+            neighbors.append(&mut self.transitive_po_neighbors(n))
+        }
+
+        neighbors
     }
 
-    pub fn potential_fences(&self) -> Vec<Fence> {
-        todo!()
+    /// Check if two nodes are connected through po+,
+    /// i.e. there is a path of [AegEdge::ProgramOrder] connecting them
+    pub fn is_po_connected(&self, a: NodeIndex, b: NodeIndex) -> bool {
+        self.transitive_po_neighbors(a).contains(&b)
+    }
+
+    pub fn tso_critical_cycles(&self) -> Vec<CriticalCycle> {
+        critical_cycles::critical_cycles(self, &Architecture::Tso)
+    }
+
+    pub fn potential_fences(&self, cycles: &[CriticalCycle]) -> Vec<EdgeIndex> {
+        critical_cycles::potential_fences(self, cycles)
```

Comment:
@Alexandru-Dumitriu here is where you could put your code for checking for potential fence placements (see `critical_cycles.rs`) but feel free to change the signature to what you think is best

Reply:
Thinking a bit more about it I think the easiest way to implement it would be to use [petgraph::algo::astar::astar](https://docs.rs/petgraph/latest/petgraph/algo/astar/fn.astar.html) on every adjacent pair in `CriticalCycle` over the aeg, while using a high cost for edges that aren't `ProgramOrder`. Then you'll get the path between those pairs which should only be program order edges, fence placement is possible between any of those edges.